tests/roots/test-resolve-typing-guard/demo_typing_guard.py

```diff
@@ -29,6 +29,15 @@ def a(f: Decimal, s: AnyStr) -> Sequence[AnyStr | Decimal]:
     return [f, s]
 
 
+class SomeClass:
+    """This class do something."""
+
+    if TYPE_CHECKING:  # Classes doesn't have `__globals__` attribute
+
+        def __getattr__(self, item: str):  # noqa: U100
```

Comment:
This is an interesting example. However, I don't think we should be evaluating these blocks; our evaluation scope is global and this is clearly tied to a class. Perhaps we should ensure that  `if TYPE_CHECKING` is not preceded by any whitespace to avoid such scooping issues.

Reply:
One downside here would be that we'd not be able to generate type information for methods defined as such. I don't think though we have a better solution here as often time the reason for using `TYPE_CHECKING` is to avoid circular dependencies, so if we'd set this to `True` we'd fail to import the module altogether. For the time being I'd consider such code unsupported by the plugin 🤔 and perhaps we should just ignore them to not generate the warning.

Reply:
But if we ignore that code we'll get more errors, `if TYPE_CHECKING` is more than avoiding circular import. It's about adapting the code to be checked correctly in the type-checking phase. It's also used with `if not TYPE_CHECKING`.

Reply:
> But if we ignore that code we'll get more errors,

Such as?

> It's also used with `if not TYPE_CHECKING`.

Haven't seen this until now, and I'd say if you're using it like that you're doing something unsupported 🤔 

We cannot evaluate such nested (method/class) blocks because we don't have access to the scope where to populate it (only for global we have easy access to know the scope). And we cannot just set type_checking to True because that might break due to circular imports. Perhaps we should have a flag to switch between:

- set type checking true for a file and don't support circular import cases
- evaluate global scope as follow-up (what we do now) and do not support nested type check flags - handles circular import problems

Reply:
> I'd say if you're using it like that you're doing something unsupported 

Yeah, that was what I meant. If `mypy`, for example, doesn't support some advanced features and we need to prevent the code to be executed in the type-checking phase!
I can change those modifications to:

```python
# CODE
if hasattr(obj, "__globals__"):
    exec(guarded_code, obj.__globals__)
# CODE
```

This way we prevent any non-module code to be executed!

Reply:
I'd still prefer to alter the regex, I'd want to not even get to this state for such code.

Reply:
yeah 🤔

Reply:
Done, I've made the tests and adding just a `\n` at the begging do the work!